File: 745.cpp
#include <chrono>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <vector>

using std::int64_t;

// Project Euler 745 scratch implementation.
//
// The target quantity is
//
//     S(n) = sum_{m=1..n} g(m)
//
// where g(m) is the largest perfect-square divisor of m.
//
// For each k >= 1, the numbers with largest square divisor exactly k^2 are
// precisely the values m = k^2 * s where s is squarefree. That lets us write
// the answer as
//
//     S(n) = sum_{k^2 <= n} k^2 * Q(floor(n / k^2))
//
// where Q(x) is the number of squarefree integers up to x.
//
// The squarefree counting step is handled with Möbius inversion:
//
//     Q(x) = sum_{d^2 <= x} mu(d) * floor(x / d^2)
//
// This file is a direct C++ translation of the same Möbius-based experiment in
// 745.py. It is much better than brute-force marking of multiples, while still
// leaving room for further algorithmic optimization at very large scales.

std::vector<int> generate_mobius(int n) {
    // Compute mu(0..n) with a linear sieve.
    //
    // mu(d) is:
    // -  1 when d has an even number of distinct prime factors
    // - -1 when d has an odd number of distinct prime factors
    // -  0 when d is divisible by any square > 1
    std::vector<int> mu(n + 1, 0);
    std::vector<bool> is_prime(n + 1, true);
    std::vector<int> primes;

    mu[1] = 1;
    if (n >= 0) {
        is_prime[0] = false;
    }
    if (n >= 1) {
        is_prime[1] = false;
    }

    for (int i = 2; i <= n; ++i) {
        if (is_prime[i]) {
            primes.push_back(i);
            mu[i] = -1;
        }

        for (int p : primes) {
            if (static_cast<int64_t>(i) * p > n) {
                break;
            }
            is_prime[i * p] = false;

            if (i % p == 0) {
                mu[i * p] = 0;
                break;
            } else {
                mu[i * p] = -mu[i];
            }
        }
    }

    return mu;
}

int64_t solve(int64_t n) {
    // Evaluate
    //
    //   sum_{k^2 <= n} k^2 * Q(floor(n / k^2))
    //
    // directly, where Q is the squarefree-counting function derived from the
    // Möbius formula above.
    static constexpr int64_t mod = 1000000007LL;
    const int64_t root = static_cast<int64_t>(std::sqrt(static_cast<long double>(n)));
    const std::vector<int> mu = generate_mobius(static_cast<int>(root));
    int64_t total = 0;

    for (int64_t k = 1; k <= root; ++k) {
        int64_t d = 1;
        int64_t cnt = 0;
        while (true) {
            const int64_t kd = k * d;
            const int64_t denom = kd * kd;
            // floor(n / (k^2 * d^2)) counts how many multiples of d^2 remain
            // after removing the k^2 factor.
            const int64_t val = n / denom;
            if (val < 1) {
                break;
            }
            cnt += static_cast<int64_t>(mu[static_cast<std::size_t>(d)]) * val;
            ++d;
        }

        // Every value in this class contributes k^2 to the final sum.
        total = (total + ((cnt % mod + mod) % mod) * ((k * k) % mod)) % mod;
    }

    return total;
}

int main() {
    // Large input used for performance experiments.
    const auto start = std::chrono::steady_clock::now();
    const int64_t n = 100000000000000LL;
    const int64_t total = solve(n);
    const auto end = std::chrono::steady_clock::now();
    const std::chrono::duration<double> elapsed = end - start;
    std::cout << "Total: " << total << '\n';
    std::cout << "Time: " << elapsed.count() << " seconds\n";
    return 0;
}

File: 745.py

"""
Project Euler 745 scratch implementation.

This file computes

    S(n) = sum_{m=1..n} g(m)

where g(m) is the largest perfect-square divisor of m.

The implementation uses the identity

    count_squarefree(x) = sum_{d^2 <= x} mu(d) * floor(x / d^2)

to count, for each square k^2, how many integers m <= n have largest square
divisor exactly k^2. The contribution of that class is then

    k^2 * count_squarefree(floor(n / k^2)).

This is a direct Möbius-based implementation intended for experimentation and
cross-checking. It is much better than the original brute-force marking
approach, but the nested summation still leaves substantial optimization
headroom for very large n.
"""

from math import isqrt
from time import perf_counter


def generate_mobius(n):
    """Return mu(0..n) using a linear sieve.

    mu(d) is the Möbius function:
    -  1 if d is a product of an even number of distinct primes
    - -1 if d is a product of an odd number of distinct primes
    -  0 if d is divisible by a square

    The squarefree-counting formula in ``solve`` only needs values up to
    floor(sqrt(n)).
    """
    mu = [0] * (n + 1)
    is_prime = [True] * (n + 1)
    primes = []

    mu[1] = 1
    is_prime[0] = is_prime[1] = False

    for i in range(2, n + 1):
        if is_prime[i]:
            primes.append(i)
            mu[i] = -1

        for p in primes:
            if i * p > n:
                break
            is_prime[i * p] = False

            if i % p == 0:
                mu[i * p] = 0
                break
            else:
                mu[i * p] = -mu[i]

    return mu


def solve(n):
    """Compute sum of largest square divisors for 1 <= m <= n, modulo 1e9+7.

    For a fixed k, define the class of numbers

        m = k^2 * s

    where s is squarefree. Exactly those numbers have largest square divisor
    equal to k^2, so the number of such m <= n is the count of squarefree
    integers up to floor(n / k^2).

    The inner loop evaluates that squarefree count with Möbius inversion:

        count_squarefree(x) = sum mu(d) * floor(x / d^2).

    This keeps the logic close to the mathematics, though it still performs a
    nested summation over k and d.
    """
    mod = 10**9 + 7
    root = isqrt(n)
    mu = generate_mobius(root)
    total = 0
    for k in range(1, root + 1):
        d = 1
        cnt = 0
        while True:
            # floor(n / (k^2 * d^2)) is the number of multiples of d^2 that
            # remain after factoring out k^2.
            val = n / ((k * d) ** 2)
            if val < 1:
                break
            cnt += mu[d] * int(val)
            d += 1

        # All numbers whose largest square divisor is exactly k^2 contribute
        # k^2 each to the final sum.
        total = (total + (cnt * (k ** 2)) % mod) % mod
    return total


if __name__ == "__main__":
    # Large input used for performance experiments.
    start = perf_counter()
    n = 10**14
    total = solve(n)
    elapsed = perf_counter() - start
    print(f"Time: {elapsed:.2f} seconds")
    print(f"Total: {total}")
